daemon/cmd/devices.go

```diff
@@ -255,22 +251,29 @@ links:
 func expandDevices() error {
 	allLinks, err := netlink.LinkList()
 	if err != nil {
-		return fmt.Errorf("Cannot list network devices via netlink: %w", err)
+		return fmt.Errorf("Device wildcard expansion failed to fetch devices: %w", err)
 	}
-	expandedDevices := make(map[string]bool)
+	expandedDevices := make(map[string]struct{})
 	for _, iface := range option.Config.Devices {
 		if strings.HasSuffix(iface, "+") {
 			prefix := strings.TrimRight(iface, "+")
 			for _, link := range allLinks {
 				attrs := link.Attrs()
 				if strings.HasPrefix(attrs.Name, prefix) {
-					expandedDevices[attrs.Name] = true
+					expandedDevices[attrs.Name] = struct{}{}
 				}
 			}
 		} else {
-			expandedDevices[iface] = true
+			expandedDevices[iface] = struct{}{}
 		}
 	}
+	if len(option.Config.Devices) > 0 && len(expandedDevices) == 0 {
+		// User defined devices, but expansion yielded no devices. Fail here to not
+		// surprise with auto-detection.
+		return fmt.Errorf("Device wildcard expansion failed to detect devices. Please verify --%s option.",
```

Comment:
I think we are changing the previous behavior (i.e., do the dev detection if the expansion has failed), which is fine, just that we need to add a note to the upgrade guide.

Reply:
Added a note about that, plus about the change in device detection logic.